File: day-rotation.js
let y = 0;
let m = 0;
let d = 0;

const widget = new ListWidget();

const start = new Date(2023, 8, 6);
const end = new Date(2024, 5, 28);
const lsYear = new Date(2023, 5, 1);
const lsYearEnd = new Date(2023, 5, 30);
let current = new Date();
if (y != 0 && m != 0 && d != 0) {
    current = new Date(y, m-1, d);
}


const holidays = new Map([
    [new Date(2023, 8, 22).getTime(), "Pro-D Day"],
    [new Date(2023, 9, 2).getTime(), "Truth and Reconciliation Day"],
    [new Date(2023, 9, 9).getTime(), "Thanksgiving Day"],
    [new Date(2023, 9, 20).getTime(), "Pro-D Day"],
    [new Date(2023, 10, 10).getTime(), "School Closure Day"],
    [new Date(2023, 10, 13).getTime(), "Remembrance Day"],
    [new Date(2023, 10, 17).getTime(), "Pro-D Day"],
    [new Date(2024, 1, 19).getTime(), "Family Day"],
    [new Date(2024, 1, 23).getTime(), "Pro-D Day"],
    [new Date(2024, 2, 29).getTime(), "Good Friday"],
    [new Date(2024, 3, 1).getTime(), "Easter Monday"],
    [new Date(2024, 3, 18).getTime(), "Pro-D Day"],
    [new Date(2024, 4, 19).getTime(), "Victoria Day"],
    [new Date(2024, 5, 6).getTime(), "Pro-D Day"]
]);

const wBreakStart = new Date(2023, 11, 23);
const wBreakEnd = new Date(2024, 0, 7);
const sBreakStart = new Date(2024, 2, 16);
const sBreakEnd = new Date(2024, 2, 28);

for (let date = new Date(wBreakStart); date <= wBreakEnd; date.setDate(date.getDate() + 1)) {
    holidays.set(new Date(date), "Winter Break");
}

for (let date = new Date(sBreakStart); date <= sBreakEnd; date.setDate(date.getDate() + 1)) {
    holidays.set(new Date(date), "Spring Break");
}

let titleText = "";
if (current.toDateString() === new Date(2023, 8, 5).toDateString()) {
    titleText += "Back to School!";
} else if (current.getTime() < new Date(2023, 8, 5).getTime()) {
    if (current.getTime() < lsYearEnd.getTime()) {
        const dayN = Math.floor((current - lsYear) / (1000 * 60 * 60 * 24)) % 2;
        if (dayN === 0) {
            titleText += "Day 1";
        } else {
            titleText += "Day 2";
        }
    } else {
        titleText += "Enjoy Summer!"
    }
} else if (holidays.has(current.getTime())) {
    titleText += holidays.get(current.getTime());
} else if (current.getDay() === 0 || current.getDay() === 6) {
    titleText += "Weekend";
} else if (current.getTime() >= end.getTime()) {
    titleText += "Enjoy Summer!";
} else {
    let totalDays = Math.floor((current - start) / (1000 * 60 * 60 * 24))
    let sum = 0;
    for (const [key, value] of holidays) {
        if (key < current) {
            sum++;
        }
    }
    totalDays -= sum;
    const dayN = totalDays % 2;
    if (dayN === 0) {
        titleText += "Day 1";
    } else {
        titleText += "Day 2";
    }
}

console.log(titleText);
console.log(current.toLocaleDateString(undefined, {year: "numeric", month: "long", day: "numeric"}));

File: day_rotation.py
import datetime

start = datetime.date(2023, 9, 6)
end = datetime.date(2024, 6, 27)
ls_year = datetime.date(2023, 6, 1)
ls_year_end = datetime.date(2023, 6, 30)
current = datetime.date.today()

holidays = {
    datetime.date(2023, 9, 22): "Pro-D Day",
    datetime.date(2023, 10, 2): "Truth and Reconciliation Day",
    datetime.date(2023, 10, 9): "Thanksgiving Day",
    datetime.date(2023, 10, 20): "Pro-D Day",
    datetime.date(2023, 11, 10): "School Closure Day",
    datetime.date(2023, 11, 13): "Remembrance Day",
    datetime.date(2023, 11, 17): "Pro-D Day",
    datetime.date(2024, 2, 19): "Family Day",
    datetime.date(2024, 2, 23): "Pro-D Day",
    datetime.date(2024, 3, 29): "Good Friday",
    datetime.date(2024, 4, 1): "Easter Monday",
    datetime.date(2024, 4, 19): "Pro-D Day",
    datetime.date(2024, 5, 20): "Victoria Day",
    datetime.date(2024, 6, 7): "Pro-D Day"
}

w_break_start = datetime.date(2023, 12, 23)
w_break_end = datetime.date(2024, 1, 7)
s_break_start = datetime.date(2024, 3, 16)
s_break_end = datetime.date(2024, 3, 28)

for i in range((w_break_end - w_break_start).days + 1): holidays[w_break_start + datetime.timedelta(days = i)] = "Winter Break"
for i in range((s_break_end - s_break_start).days + 1): holidays[s_break_start + datetime.timedelta(days = i)] = "Spring Break"

print(current)

if current == datetime.date(2023, 9, 5): print("Back to School!")
elif current < datetime.date(2023, 9, 5):
    if current < ls_year_end:
        day_n = (current - start).days % 2
        print("Day 1" if day_n == 0 else "Day 2")
    else: print("Enjoy Summer!")
elif current in holidays: print(holidays[current])
elif current.weekday() >= 5: print("Weekend")
elif current > end: print("Have a Good Summer!")
else:
    total_days = (current - start).days - sum(1 for holiday in holidays if holiday < current)
    day_n = total_days % 2
    print("Day 1" if day_n == 0 else "Day 2")
